Add initial support for connected datagram sockets

simonjbeaumont opened the pull request:
### Motivation:

From the man page for `connect(2)`:
>     The parameter socket is a socket.  If it is of type SOCK_DGRAM, this call specifies the peer
>     with which the socket is to be associated; this address is that to which datagrams are to be
>     sent, and the only address from which datagrams are to be received.

This presents an opportunity for some performance optimisations for applications that want to use UDP with a fixed peer, but is currently prohibited in NIO.

This PR adds some preliminary support for connected datagram sockets and allow further work to optimise that code path.

### Modifications:

* Allow the calling of `sendmsg(2)` with `NULL` for `msg_name` by changing the `Socket` types.
* Keep track of whether the `DatagramChannel` is connected in `PendingDatagramWritesManager` and, if it is connected, use `NULL` for `msg_name` in the headers for scalar and vector writes.
* Allow connect to be called after bind in `BaseSocketChannel`.
* Allow `connectSocket(to:)` to be called on `DatagramChannel`.
* Rename `DatagramBootstrap.bind0(makeChannel:registerAndBind:)` to `withNewChannel(makeChannel:bringup:)` to account for the fact that it isn't only used for `bind`.
* Add set of `DatagramBootstrap.connect(...)` APIs.
* Remove DatagramChannelTests.testConnectionFails because that doesn't fail anymore.
* Add ConnectedDatagramChannelTests, inheriting from DatagramChannelTests but connecting both channels together.
* Use connected-mode UDP in NIOUDPEchoClient.

### Result:

* Can now call `Channel.connect(to:)` when using datagram sockets.
* Some minor performance improvements in `PendingDatagramWritesManager` because the address is no longer used (or copied).
* Opportunities for future performance enhancements.
* All tests pass on Darwin.

### Considerations:

* Should we add support for `disconnectx(2)`?
* To support `connect` after `bind` the state machine in `BaseSocketChannel` was made more lenient. Is this acceptable? 
* Is the `NIOUDPEchoClient` considered API? If not, can we rework the argument parsing so we can include a `--connect` flag?
* Need to make sure we're testing the codepaths on Windows (different socket API) and Linux (which uses `sendmmsg` for vector writes).

Lukasa commented on Sources/NIOPosix/Bootstrap.swift:
```diff
@@ -907,14 +907,82 @@ public final class DatagramBootstrap {
+    /// - parameters:
+    ///     - unixDomainSocketPath: The path of the UNIX Domain Socket to connect to. `path` must not exist, it will be created by the system.
+    public func connect(unixDomainSocketPath: String) -> EventLoopFuture<Channel> {
+        return bind0 {
```
Should this still call `bind0`? (This suggests a missing test)

simonjbeaumont replied:
Um, no. Copy-pasta error here. This should clearly be `connect0`. Addressed in af93fbea7.

Just checked and none of the existing `DatagramChannelTests` used anything other than `bind(host:port:)` and so the `ConnectedDatagramChannelTests` also don't use this path.

Lukasa commented on Sources/NIOPosix/Bootstrap.swift:
```diff
@@ -907,14 +907,82 @@ public final class DatagramBootstrap {
+    /// - parameters:
+    ///     - unixDomainSocketPath: The path of the UNIX Domain Socket to connect to. `path` must not exist, it will be created by the system.
+    ///     - cleanupExistingSocketFile: Whether to cleanup an existing socket file at `path`.
+    public func connect(unixDomainSocketPath: String, cleanupExistingSocketFile: Bool) -> EventLoopFuture<Channel> {
```
I don't think `cleanupExistingSocketFile` makes sense on a `connect` call, so we don't need this path.

simonjbeaumont replied:
Removed in af93fbea7.

Lukasa commented on Sources/NIOPosix/PendingDatagramWritesManager.swift:
```diff
@@ -515,14 +541,28 @@ final class PendingDatagramWritesManager: PendingWritesManager {
-                try pending.data.withUnsafeReadableBytes {
-                    try scalarWriteOperation($0, addrPtr, socklen_t(addrSize), pending.metadata)
+            let writeResult: IOResult<Int>
+            if self.state.isConnected {
```
I think in this case we want to confirm that any write we're about to send has an address that matches the address we connected to.

simonjbeaumont replied:
Oh, I was thinking that this would be an opportunity to get a small perf win. I realise there'll be a perf win within the call to `sendmsg`, but by not bothering with the `withSockAddr` here we avoid a copy. 

This comes down to a question about the layer of abstraction we want to provide here I suppose. Do we want to provide an API that checks you are doing the right thing or be more transparent?

I guess we are already abstracting `sendmsg` calls away because calling them with a `msg_name` will cause an error.

WDYT about checking that the address matches on the way _in_ to PDWM, in the call to `add` if it currently `isConnected`? This would make the actual writing out of data quicker?

Now we're talking about this, it's going to be even more complicated/costly in the vector write case to make this check for each of the message headers.

Do you think there's any milage in an API that, once you call `connect()` sends all datagrams to that address? This would be the most performant, but I wonder if it might mean we should be exposing different APIs for sending these messages (that don't take an `AddressedEnvelope`) and disallow the `AddressedEnvelope` calls when connected?

simonjbeaumont replied:
Is a4ee0756d what you had in mind?

Lukasa replied:
@simonjbeaumont Yeah the performance win is valuable, but I think we need to detect misuse of the API as well.

If we wanted to get the performance win I think we get it by allowing sending naked byte buffers instead of addressedenvelopes once we connect. If users still send addressed envelopes then we still have to check. I'm fine with us doing the check on the way into PDWM, but if we do that we also have to validate anything that's buffered when the `connect` call comes in.

simonjbeaumont replied:
Hm, the presence of buffering makes this an interesting problem since it's possible the connected state of the socket can change while things are buffered. That's the case even with what's in this PR, but becomes even more apparent if we start supporting `disconnectx` and re-`connect`.

simonjbeaumont replied:
OK so I reverted the commit that checks the writes when flushing and replaced it with ae1b8d409 which fails any buffered writes when connecting or reconnecting the datagram channel and validates the destination address on the way in, when the channel is connected.

Also added some tests.

Lukasa commented on Sources/NIOUDPEchoClient/main.swift:
```diff
@@ -126,14 +126,23 @@ defer {
-        return try bootstrap.bind(unixDomainSocketPath: listeningPath).wait()
+
+let channel: Channel
+let connectedMode = true
```
Currently this is unconditionally `true`: did you want to add support for both?

simonjbeaumont replied:
Yeah, that's what I meant by this in the considerations in the PR description:

> Is the NIOUDPEchoClient considered API? If not, can we rework the argument parsing so we can include a --connect flag?

If it's not considered a breaking change then I can add an option. Adding an option with the existing argument parsing strategy is gonna be pretty clunky.

Lukasa replied:
We can rewrite the argument parsing here.

simonjbeaumont replied:
Is depending on `swift-argument-parser` a no-go?

simonjbeaumont replied:
Added commit:

```
* 4d5652cc6 NIOUDPEchoClient: Connect socket to remote only if --connect is used
```

Lukasa commented on Tests/NIOPosixTests/DatagramChannelTests.swift:
```diff
@@ -912,3 +906,11 @@ final class DatagramChannelTests: XCTestCase {
     }
 }
+
+final class ConnectedDatagramChannelTests: DatagramChannelTests {
```
I think this will stop working once we adopt test discovery on Linux, which strictly works structurally and cannot traverse into subclasses. Can we either use copy-paste or have two separate classes which both defer to a common implementation?

simonjbeaumont replied:
Oh, that's disappointing. I didn't realise that.

`grpc-swift` makes use of subclasses, in that every test inherits from `GRPCTestCase` and there is further subclassing for shared tests IIRC. It also makes use of test discovery on Linux, given it's `LinuxMain.swift` consists only of:

```swift
fatalError("Use --enable-test-discovery to run tests on Linux")
```

What am I missing that means we can't do the same in NIO?

Lukasa replied:
We should double-check that all of grpc-swift's tests are actually running on Linux: my experiments suggest the answer is no.

Lukasa replied:
Looking into this:

- `GRPCTestCase` adds functionality but does not define tests, so that parent class is safe.
- `EchoTestCaseBase` adds functionality but does not define tests, so that parent class is safe.
- `FunctionalTestsInsecureTransport` defines tests, which is sketchy
- `GRPCInsecureInteroperabilityTests` defines tests, which is sketchy
- `ServerThrowingTests` defines tests, which is sketchy

Testing this locally I can confirm that the `FunctionalTestsInsecureTransport` subclasses `FunctionalTestsAnonymousClient` and `FunctionalTestsMutualAuthentication` do not run. The `GRPCInsecureInteroperabilityTests` subclass `GRPCSecureInteroperabilityTests` do not run. The `ServerThrowingTests` subclasses `ServerDelayedThrowingTests`, `ClientThrowingWhenServerReturningErrorTests`, `ServerErrorTransformingTests` do not run. This is a bug against grpc-swift.

Lukasa replied:
Filed as grpc/grpc-swift#1392.

simonjbeaumont replied:
I've added tests for the connected support to the existing class for now. I'll resolve this thread and if it turns out that we want to have separate suite, I'll make sure I don't make the mistake of subclassing :)

Lukasa commented on Sources/NIOPosix/BaseSocketChannel.swift:
```diff
@@ -140,12 +140,18 @@ private struct SocketChannelLifecycleManager {
 
+        // TODO: This is to support connect after bind for connected-mode datagram sockets. Might need a better solution for this.
+        // origin: .activated
+        case (.activated, .activate):
```
Yeah we may want to rethink this a bit. I wonder if we can construct a system whereby only datagram channel allows this.

simonjbeaumont replied:
Yes, this was what I was thinking we might be wanting.

Additionally, we should consider whether we want to support calling `connect(2)` after `connect(2)` which seems to be a valid use case to re-associate a connected datagram socket with a new address.

Lukasa replied:
Yeah we should probably allow that.

simonjbeaumont replied:
This is now handled in ae1b8d409 with some tests.

simonjbeaumont commented on Sources/NIOPosix/Bootstrap.swift:
```diff
@@ -907,14 +907,82 @@ public final class DatagramBootstrap {
+    ///     - host: The host to connect to.
+    ///     - port: The port to connect to.
+    ///
+    /// - NOTE: This function is specifically for connected-mode UDP and does not call `bind`.
```
Ah, yes. These are leftover from before I allowed connect after bind. I'll get rid of them.

Addressed in af93fbea7.

simonjbeaumont commented on Sources/NIOPosix/Bootstrap.swift:
```diff
@@ -907,14 +907,82 @@ public final class DatagramBootstrap {
+    /// - parameters:
+    ///     - address: The `SocketAddress` to connect to.
+    ///
+    /// - NOTE: This function is specifically for connected-mode UDP and does not call `bind`.
```
Addressed in af93fbea7.

simonjbeaumont commented on Sources/NIOPosix/SocketChannel.swift:
```diff
@@ -672,6 +684,14 @@ final class DatagramChannel: BaseSocketChannel<Socket> {
         let data = self.unwrapData(data, as: AddressedEnvelope<ByteBuffer>.self)
 
+        // If the datagram socket is connected then we check here that the remote is appropriate.
+        if let remoteAddress = self.remoteAddress {
+            guard data.remoteAddress == remoteAddress else {
+                promise?.fail(IOError(errnoCode: EISCONN, reason: "Socket was connected before flushing pending write."))
+                return
+            }
+        }
```
@Lukasa I'm thinking that we might get even more perf win here once PDWM supports `ByteBuffer`. In this branch we could buffer only the buffer, without the envelope. Then we wouldn't even need to do the address stripping on the way out in PDWM flush?

simonjbeaumont replied:
👍 Is that something we want in the first PR or can we do this incrementally?

simonjbeaumont replied:
OK, great. I added support for sending unaddressed `ByteBuffer` and have added tests showing how users should only use `AddressedEnvelope` with unconnected channels and `ByteBuffer` with connected channels.

simonjbeaumont commented on Sources/NIOPosix/SocketChannel.swift:
```diff
@@ -526,12 +526,24 @@ final class DatagramChannel: BaseSocketChannel<Socket> {
-        // For now we don't support operating in connected mode for datagram channels.
-        throw ChannelError.operationUnsupported
+        // TODO: this could be a channel option to do other things instead here, e.g. fail the connect
+        if !self.pendingWrites.isEmpty {
```
@Lukasa Do I need to consider `self.pendingWrites.isPendingFlush` here? I guess if that's `true` then it means that the API has been used correctly. Or do we still leave it to the user to ensure that the flush has taken place, by means of the associated future?

simonjbeaumont commented:
@Lukasa I added the `ByteBuffer` based write APIs which take a faster path and added some more testing. I think this is ready for another pass when you have time.

Lukasa commented on Sources/NIOPosix/SocketChannel.swift:
```diff
@@ -668,9 +680,44 @@ final class DatagramChannel: BaseSocketChannel<Socket> {
+    private func bufferPendingAddressedWrite(data: AddressedEnvelope<ByteBuffer>, promise: EventLoopPromise<Void>?) {
+        // It is only appropriate to use an AddressedEnvelope if the socket is not connected.
+        guard self.remoteAddress == nil else {
+            promise?.fail(IOError(errnoCode: EISCONN, reason: """
```
Here we should only error if the remote address doesn't match the address in the envelope, but otherwise it's totally fine.

simonjbeaumont replied:
Resolved by https://github.com/apple/swift-nio/pull/2084/commits/eb580d69ca04991d91b7324fee725ca25bce9af9.

Lukasa reviewed (commented):
Great stuff! I've skipped reviewing the tests for now while I have feedback in the product code to speed up our turnaround time.

simonjbeaumont commented on Sources/NIOPosix/PendingDatagramWritesManager.swift:
```diff
@@ -433,6 +463,27 @@ final class PendingDatagramWritesManager: PendingWritesManager {
+                                address: nil,
+                                metadata: nil))
+
+        if self.state.bytes > waterMark.high && channelWritabilityFlag.compareAndExchange(expected: true, desired: false) {
```
Done in 501cac07a.

simonjbeaumont reviewed (commented):
Thanks for the latest comments, @Lukasa!

I've refactored the common logic out of the two `PDWM.add(...)` functions.

I have some thoughts regarding the API semantics which I've added in https://github.com/apple/swift-nio/pull/2084/files#r858949312 that might influence the next steps.

Lukasa commented on Sources/NIOPosix/PendingDatagramWritesManager.swift:
```diff
@@ -515,14 +566,31 @@ final class PendingDatagramWritesManager: PendingWritesManager {
+            let writeResult: IOResult<Int>
+
+            if let address = pending.address {
+                assert(!self.state.isConnected, "Pending write with address on connected socket.")
```
So the problem with reflecting that semantics is the issue of `AddressedEnvelope.Metadata`. This is still meaningful to apply even in connected mode, but it's not accessible to users anymore if they connect. I don't think that's an intended outcome. Supporting both and just fixing it up allows users to wish to to keep sending the metadata but still gives access to a fast path. Does that make sense?

I don't think this conflicts with late-binding the address though: the `DatagramChannel` just needs to communicate the metadata to the PDWM, at which point it can safely drop the address before storing into the `PendingDatagramWrite`. Does that make sense?

Lukasa replied:
Also you're right about the vector case, my bad, I just missed it.

simonjbeaumont replied:
Your statements make sense. I thought, however, we wanted to offer APIs in terms of `ByteBuffer` and if we offer such APIs we should guide/force users to use the appropriate one.

Right, the point about the metadata is interesting and I came across it when I was making changes to PDWM. The additional flags passed to the `sendmsg` are a function of not just the `AddressedEnvelope.metadada` but also the `AddressedEnvelope.address.protocol`. I played around with having this property lifted into the `PendingDatagramWrite` so that it can be used even when `AddressedEnvelope.address` is `nil`.

I think if we want to continue to support these additional bits of metadata along with the write, which I completely expect that we do, we have (at least) two options:

1. `DatagramChannel.writeToSocket(_:)` accepts just an `AddressedEnvelope` and the implementation (either channel, PDWM, or both) discards the address along the way. This is an implicit API which might be harder to reason about.
2. Expose a more  explicit API that takes `ByteBuffer` and expect users to use it when connected. This API would need some way of also passing in the metadata that is appropriate, i.e. the congestion control mechanism currently composed of `AddressedEnvelope.metadata` and `AddressedEnvelope.address.protocol`, without passing the the metadata that is inappropriate, i.e. the destination address.

Lukasa replied:
Nope, my preference is that `PendingDatagramWrite` has a `nil` address when we don't want it, and then just restructure how we construct the metadata to use the connected address from `DatagramChannel` to populate the `protocol` for the metadata early.

As to the broader question of forcing folks to use the "right" API, I think unless we're willing to change the write to be "ByteBuffer + optional metadata" (which I think is not great) then it's probably fine to just support `AddressedEnvelope` for now in cases where folks want to send ECN.

simonjbeaumont replied:
OK, I relaxed support to allow for `AddressedEnvelope` in writes to a connected `DatagramChannel` in eb580d69c. It validates that the address in the envelope matches that of the connected socket.

Lukasa commented on Sources/NIOPosix/SocketChannel.swift:
```diff
@@ -526,12 +526,24 @@ final class DatagramChannel: BaseSocketChannel<Socket> {
+        // TODO: this could be a channel option to do other things instead here, e.g. fail the connect
+        if !self.pendingWrites.isEmpty {
+            self.pendingWrites.failAll(
+                error: IOError(
```
I think a NIO-specific error adds value because it helps us debug the exact cause of the problem. IMO, NIO shouldn't pretend that it isn't there when it is doing things that are unique and specific to it. In this case, we're making the decision to drop some data, and we should own that instead of pretending that the OS failed them.

simonjbeaumont replied:
That reasoning makes total sense; I'll work on a more explicit error for this case where buffered writes are failed.

simonjbeaumont commented:
@Lukasa LMK if you have any additional feedback on this one.

Lukasa commented on Sources/NIOPosix/BaseSocketChannel.swift:
```diff
@@ -140,12 +140,18 @@ private struct SocketChannelLifecycleManager {
                 pipeline.syncOperations.fireChannelUnregistered()
             }
 
+        // TODO: This is to support connect after bind for connected-mode datagram sockets. Might need a better solution for this.
```
Does this TODO still apply?

simonjbeaumont replied:
Yes, currently it is required. Without it, we get the following error.

```
NIOPosix/BaseSocketChannel.swift:166: Fatal error: illegal transition: state=activated, event=activate
2022-05-23 11:07:35.867416+0100 xctest[6470:11648282] NIOPosix/BaseSocketChannel.swift:166: Fatal error: illegal transition: state=activated, event=activate
```

I suppose, if we want to only allow this on `DatagramChannel` then we would need to change the implementation of `BaseSocketChannel.moveState(event:)` (currently private) to `fatalError("must be overridden")` and provide different implementations in `BaseStreamSocket` and `DatagramChannel`.

WDYT?
